lib/Sema/CSSimplify.cpp

```diff
@@ -4254,6 +4253,133 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+  return nullptr;
+}
+
+/// Looks up and returns the @dynamicCallable required methods (if they exist)
+/// implemented by a type. This function should not be called directly: instead,
+/// call `getDynamicCallableMethods` which performs caching.
+static DynamicCallableMethods
+lookupDynamicCallableMethods(Type type, ConstraintSystem &CS,
+                             const ConstraintLocatorBuilder &locator,
+                             bool &error) {
+  auto &ctx = CS.getASTContext();
+  DynamicCallableMethods methods;
+  methods.argumentsMethod =
+    lookupDynamicCallableMethod(type, CS, locator, ctx.Id_withArguments,
+                                /*hasKeywordArgs*/ false, error);
+  methods.keywordArgumentsMethod =
+    lookupDynamicCallableMethod(type, CS, locator, ctx.Id_withKeywordArguments,
+                                /*hasKeywordArgs*/ true, error);
+  return methods;
+}
+
+/// Returns the @dynamicCallable required methods (if they exist) implemented by
+/// a type.
+/// This function may be slow for deep class hierarchies and multiple protocol
+/// conformances, but it is invoked only after other constraint simplification
+/// rules fail.
+static DynamicCallableMethods
+getDynamicCallableMethods(Type type, ConstraintSystem &CS,
+                          const ConstraintLocatorBuilder &locator,
+                          bool &error) {
+  auto canType = type->getCanonicalType();
+  auto it = CS.DynamicCallableCache.find(canType);
+  if (it != CS.DynamicCallableCache.end()) return it->second;
+
+  auto calculate = [&]() -> DynamicCallableMethods {
+    // If this is a protocol composition, check if any of the protocols have the
+    // attribute.
+    if (auto protocolComp = canType->getAs<ProtocolCompositionType>()) {
+      DynamicCallableMethods methods;
+      for (auto protocolType : protocolComp->getMembers()) {
+        auto tmp = getDynamicCallableMethods(protocolType, CS, locator, error);
+        if (error) return methods;
+        if (tmp.argumentsMethod) {
+          if (methods.argumentsMethod) {
+            error = true;
+            return methods;
+          }
+          methods.argumentsMethod = tmp.argumentsMethod;
+        }
+        if (tmp.keywordArgumentsMethod) {
+          if (methods.keywordArgumentsMethod) {
+            error = true;
+            return methods;
+          }
+          methods.keywordArgumentsMethod = tmp.keywordArgumentsMethod;
+        }
+      }
+      return methods;
+    }
+
+    // Otherwise, this must be a nominal type.
+    // Dynamic calling doesn't work for tuples, etc.
+    auto nominal = canType->getAnyNominal();
+    if (!nominal) return DynamicCallableMethods();
+
+    // If this type conforms to a protocol which has the attribute, then
+    // look up the methods.
+    for (auto p : nominal->getAllProtocols())
+      if (p->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
```

Comment:
What about protocols that refine other protocols?

Reply:
There's a protocol refinement test [here](https://github.com/dan-zheng/swift/blob/fe91f19de1c273e6179aaa08a71fa145ae314d41/test/NameBinding/dynamic-callable.swift#L197).
I'd be happy to add more tests, if you have suggestions.